### ISSUE TITLE
Add support for updating gradle wrapper

### ISSUE THREAD

**Issue by rahulsom** [dependabot/dependabot-core#2223]
Hi!

This is an amazing way to manage dependencies.

It would be awesome if it could also send PRs for gradle wrapper updates. 
Typically this involves 

1. Find the latest version of gradle published at 

```
https://api.github.com/repos/gradle/gradle/releases/latest
```

2. Run 

```
./gradlew wrapper --gradle-version ${VERSION} --distribution-type all
```

3. Send a PR.

Some projects using maven also use a wrapper - https://github.com/takari/maven-wrapper but that can be a separate issue.

**Comment by FireMasterK** [comment]
Any progress on this?

**Comment by robstoll** [comment]
@infin8x do you have some pointers where this needs to be implemented, I would still take a look

**Comment by infin8x** [comment]
@robstoll I don't at the moment, sorry. We're a bit swamped at the moment and can't commit to giving your PR the requisite attention.

**Comment by robstoll** [comment]
@infin8x I don't need a commitment, just a pointer where I should start looking. There is a lot of code and I don't have time either, so if I spend time helping you out then I want to do it as efficiently as possible. I think it would be enough, if you can point me to the file where the analysis for updates takes place. Thanks in advance

**Comment by JavierSegoviaCordoba** [comment]
Poke, should be great someone who knows the project can help @robstoll to get this working.

**Comment by sehenst** [comment]
> Poke, should be great someone who knows the project can help @robstoll to get this working.

@asciimike ?

**Comment by jameswald** [comment]
I have been using https://github.com/marketplace/actions/update-gradle-wrapper-action to implement this feature. Gets the job done for now although it would be better if this were supported by dependabot.

**Comment by asciimike** [comment]
https://github.com/dependabot/dependabot-core#dependabot-package-manager is a good place to start, as it has a good explanation of how the various pieces fit together, and links to READMEs on how to implement the required methods.

https://github.com/dependabot/dependabot-core/tree/main/gradle/lib/dependabot/gradle is probably where most of the additions will happen. I'd expect you'd have to modify the update checker as well as the file updater.

**Comment by robstoll** [comment]
I am afraid,but I don't have the time any more to look at it but it hopefully helps someone else

**Comment by IdiosApps** [comment]
I've been reading about this and have opened https://github.com/IdiosApps/dependabot-gradlewrapper-test. It gives some context/summary of the problem, and a few gradle dependencies so we can test any changes to dependabot-core with a dry run on it.

I'll try to find time to look into this repo later, thanks for the links @asciimike

**Comment by mustafaozhan** [comment]
Really looking forward to seeing this too!

**Comment by eirnym** [comment]
I'm looking forward to support Gradle Wrapper updates. Update-gradle-wrapper-action mentioned above has too many problems. I'm going to replace it using https://services.gradle.org/ at least till the dependabot will support it

**Comment by eirnym** [comment]
https://github.com/marketplace/actions/upgrade-gradle is the closest implementation to what I have locally

**Comment by xtermi2** [comment]
Really disappointing, no progress here. I'm waiting for years for this feature.

**Comment by GregJohnStewart** [comment]
Bump!

Sad to see it's not supported already :(

**Comment by FunkyMuse** [comment]
any hope on this?

**Comment by deivid-rodriguez** [comment]
Hello! We'll make sure to update this issue once there's movement. In the meantime, please use reactions instead of [+1 comments](https://github.com/dependabot/dependabot-core#no-1-comments) to show support for the request. Thanks!

**Comment by DPUkyle** [comment]
https://services.gradle.org/versions/current is a great endpoint with easy-to-consume json.

**Comment by britter** [comment]
@DPUkyle I think the main problem here is that in order to upgrade the wrapper jar and shell scripts completely one needs to run the Gradle build. I'm not sure that is supported by dependabot.

**Comment by Thunderforge** [comment]
Aren't the wrapper jar and shell scripts known quantities for each release, used the same by everyone? Is it possible to have Dependabot just supply pre-built for the requested version instead of running `gradlew wrapper` to generate them?

**Comment by gmazzo** [comment]
This has been open for several years now, and since I'm tired of using 3rd party stuff for this kind of bump, I took some time to implement it #12891

I just hope that maintainers will show some ❤️ to it, since this is my first time with a big-scale Ruby project

**Comment by huehnerlady** [comment]
@rahulsom thank you so much for implementing this. is there anything special I need to do for it to be enabled? My file looks like this:

```yaml
version: 2
updates:
  - package-ecosystem: "github-actions"
    directory: "/"
    schedule:
      interval: "daily"
  - package-ecosystem: "gradle"
    directory: "/"
    schedule:
      interval: "daily"
    open-pull-requests-limit: 10
    groups:
      spring: # always update spring dependencies together
        patterns:
          - "org.springframework*"
      other: # group all non-spring minor/patch updates
        patterns:
          - "*"
        exclude-patterns:
          - "org.springframework*"
        update-types:
          - minor
          - patch
```

**Comment by yeikel** [comment]
@huehnerlady Your configuration is correct, however, there is still work in progress for this. There is a chance that we closed this issue a bit prematurely/optimistically. 

See https://github.com/dependabot/dependabot-core/pull/12891 for updates

**Comment by huehnerlady** [comment]
@yeikel ok thank you. 

Maybe it makes sense to reopen it again to have transparency on this? :)

**Comment by yeikel** [comment]
@kbukum1 Could you please re-open this issue. We should close this after we fix all the open issues and promote it from experiment to stable

Otherwise, it causes confusion like what @huehnerlady observed

**Comment by v-kbukum1** [comment]
> [@kbukum1](https://github.com/kbukum1) Could you please re-open this issue. We should close this after we fix all the open issues and promote it from experiment to stable
> 
> Otherwise, it causes confusion like what [@huehnerlady](https://github.com/huehnerlady) observed

Since there are a few issues @gmazzo and @yeikel are working on, we are going to keep this issue until we roll out the feature flag for gradle wrapper update.

**Comment by gregjotau** [comment]
Are there any docs for how to enable this? Our current dependabot.yml:
```
version: 2
updates:
  - package-ecosystem: "gradle"
    directories:
      - "/"
      - "/*"        # all immediate subfolders
    schedule:
      interval: "cron"
      cronjob: "0 0 1 * *"   # 1st of every month 00:00 UTC
    groups:
      all-gradle-updates:
        patterns: ["*"]
  - package-ecosystem: "npm"
    directories:
      - "/"
      - "/cloudflare-worker"
    schedule:
      interval: "cron"
      cronjob: "0 0 1 * *"   # 1st of every month 00:00 UTC
    groups:
      all-npm-updates:
        patterns: ["*"]
```

**Comment by yeikel** [comment]
Your configuration would work as is. 

The reason you do not see any pull request yet is because this is still in progress with limited rollout. 

You can expect that it will be ready for general availability when https://github.com/github/docs/pull/39954 is merged and this issue is also closed. 

Until then, it will be enabled on limited capacity basis for some repos until we have confidence for a general availability release

**Comment by v-kbukum1** [comment]
Hi team, sorry for being late. Support Gradle Wrapper feature flag rolled out 100%. I am planing to monitor for the day. Let me know if you get any issue. Also I will update the documentation accordingly.

**Comment by jameswald** [comment]
This is working great for me. Here's a potential enhancement that may be worth considering to ensure the integrity of the wrapper jar: https://github.com/dependabot/dependabot-core/pull/12891#issuecomment-3793026602

**Comment by mkuthan** [comment]
How can I disable updates for gradle wrapper and keep updates for regular libraries? How to define `dependency-name` in the `ignore` section?

**Comment by StefMa** [comment]
> How can I disable updates for gradle wrapper and keep updates for regular libraries? How to define `dependency-name` in the `ignore` section?

The name should be `gradle-wrapper`.

**Comment by marcindabrowski** [comment]
It is working! Thanks!

But found issue with networkTimeout property: https://github.com/dependabot/dependabot-core/issues/14024

**Comment by mkutz** [comment]
I saw this working on one of my projects, but the `gradlew` file was missing the executable flag.

**Comment by hfhbd** [comment]
@kbukum1 What about GitHub Enterprise Cloud with data residency, do you still need some flags or a new release? The dependabot run uses the latest `ghcr.io/dependabot/dependabot-updater-gradle` image but it does not create a Gradle wrapper PR.

**Comment by gmazzo** [comment]
> [@kbukum1](https://github.com/kbukum1) What about GitHub Enterprise Cloud with data residency, do you still need some flags or a new release? The dependabot run uses the latest `ghcr.io/dependabot/dependabot-updater-gradle` image but it does not create a Gradle wrapper PR.

The feature is behind `gradle_wrapper_updater` feature flag. Probably it has to be enabled for you as well.

**Comment by gmazzo** [comment]
> I saw this working on one of my projects, but the `gradlew` file was missing the executable flag.

This is being discussed at #14027

**Comment by yeikel** [comment]
> > [@kbukum1](https://github.com/kbukum1) What about GitHub Enterprise Cloud with data residency, do you still need some flags or a new release? The dependabot run uses the latest `ghcr.io/dependabot/dependabot-updater-gradle` image but it does not create a Gradle wrapper PR.

Let's wait on @kbukum1 / team. But we can probably tell from the logs in the meantime. What is the digest of `ghcr.io/dependabot/dependabot-updater-gradle` that is running in your environment? 

Also, for the flag, do you see the flag on in your logs? When the job runs, you should see `"gradle-wrapper-updater":true`  under `Job definition:` 

> The feature is behind `gradle_wrapper_updater` feature flag. Probably it has to be enabled for you as well.

This seems to be enabled globally right now(if you create a new repo, it is ON) but it is possible that rollout in enterprise/other environments may be different. I believe that we may be able to send a change to remove the flag altogether at some point

**Comment by hfhbd** [comment]
@yeikel

> What is the digest of ghcr.io/dependabot/dependabot-updater-gradle that is running in your environment?

Like I said, the latest one (at the time): `1dd17b7c49d7d8a1e1b3563fb8a96320223feb1b`

> you should see "gradle-wrapper-updater":true under Job definition:

Nope, that property is not part of the job definition.

**Comment by yeikel** [comment]
> [@yeikel](https://github.com/yeikel)
> 
> > What is the digest of ghcr.io/dependabot/dependabot-updater-gradle that is running in your environment?
> 
> Like I said, the latest one (at the time): `1dd17b7c49d7d8a1e1b3563fb8a96320223feb1b`
> 
I am sorry, I missed that in the discussion. That looks like a tag tough, not a digest (or at least, it does not resolve from my side)

That tag is currently pointing to `v0.359.0` which is the latest and includes the changes.

> 
> Nope, that property is not part of the job definition.

Then, that explains it. Although it is currently in "global rollout" it still needs the flag. it may be different for that environment where you run and we need to wait for the Dependabot's team to confirm why is not enabled there yet

**Comment by v-kbukum1** [comment]
> [@kbukum1](https://github.com/kbukum1) What about GitHub Enterprise Cloud with data residency, do you still need some flags or a new release? The dependabot run uses the latest `ghcr.io/dependabot/dependabot-updater-gradle` image but it does not create a Gradle wrapper PR.

For this currently it is behind feature flag. We need to make a plan to cleanup the feature flag and backport for Github Enterprise. I will let you know when we are doing that. 

CC: @honeyankit

**Comment by yeikel** [comment]
For full Github Enterprise support we may also need #13539

**Comment by lnhrdt** [comment]
@yeikel thanks for the updates! Regarding the GHES roadmap, is this feature slated for the next minor release? While #13539 would be a great addition, our teams on GHES are primarily using public registries and would benefit from the wrapper updates even before that enhancement is finalized.

**Comment by yeikel** [comment]
> our teams on GHES are primarily using public registries

Thanks for the insights, if that's the current use base maybe it'll be fine to start without this.

I was coming from my own anecdotal experience. In our case, public access is a non-starter since we mirror all repositories, and I’d guess other orgs may have similar constraints. That said, we’re not officially using Dependabot yet(we use core) so maybe that's why we don't matter yet. 

cc @marcindabrowski Could you share your thoughts here? Based on your recent activity, it looks like you might have more accurate context and/or a need for private registry support.

**Comment by v-kbukum1** [comment]
For GHES/Enterprise versions we are now working on plan to make the feature available. We will let you know when it is ready and for which GHES release it will be available.

**Comment by joffrey-bion** [comment]
Thanks a lot for this long-awaited feature!

I just wish this was a separate ecosystem so I can put different labels for the wrapper update PRs. These are not the same kinds of updates (dependencies VS tooling).

I guess there might be a workaround by declaring the `gradle` ecosystem twice and ignoring `gradle-wrapper` on one side, and allowing only `gradle-wrapper` on the other side, but that's pretty tedious.

**Comment by yeikel** [comment]
> Thanks a lot for this long-awaited feature!
> 
> I just wish this was a separate ecosystem so I can put different labels for the wrapper update PRs. These are not the same kinds of updates (dependencies VS tooling).
> 
> I guess there might be a workaround by declaring the `gradle` ecosystem twice and ignoring `gradle-wrapper` on one side, and allowing only `gradle-wrapper` on the other side, but that's pretty tedious.

This is what I'd do
 
- Listen on pull requests from `dependabot[bot]`
- If the title contains `gradle-wrapper` -> Add additional tags 

This would not work for grouped updates but based on your description, it seems that you'd be treating it separately 

The other alternative you have is to look at the updated files. All the gradle wrapper updates touch the same files
